providers/fips/self_test_kats.c

```diff
@@ -647,6 +647,222 @@ static int self_test_asym_cipher(const ST_KAT_ASYM_CIPHER *t, OSSL_SELF_TEST *st
+
+    BIO_snprintf(buf, sizeof(buf), "%s %s", t->desc, test);
+    return buf;
+}
+
+/*
+ * FIPS 140-3 IG 10.3.A resolution 14 mandates a CAST for ML-KEM
+ * encapsulation.
+ */
+static int self_test_kem_encapsulate(const ST_KAT_KEM *t, OSSL_SELF_TEST *st,
+                                     OSSL_LIB_CTX *libctx, EVP_PKEY *pkey)
+{
+    int ret = 0;
+    EVP_PKEY_CTX *ctx;
+    unsigned char *wrapped = NULL, *secret = NULL;
+    size_t wrappedlen, secretlen;
+    OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+
+    OSSL_SELF_TEST_onbegin(st, OSSL_SELF_TEST_TYPE_KAT_KEM,
+                           self_test_kem_description(t, "encap"));
+
+    ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, "");
+    if (ctx == NULL)
+        goto err;
+
+    *params = OSSL_PARAM_construct_octet_string(OSSL_KEM_PARAM_IKME,
+                                                (unsigned char *)t->entropy,
+                                                t->entropy_len);
+    if (EVP_PKEY_encapsulate_init(ctx, params) <= 0)
+        goto err;
+
+    /* Get output sizes */
+    if (EVP_PKEY_encapsulate(ctx, NULL, &wrappedlen, NULL, &secretlen) <= 0)
+        goto err;
+    wrapped = OPENSSL_malloc(wrappedlen);
+    secret = OPENSSL_malloc(secretlen);
+    if (wrapped == NULL || secret == NULL)
+        goto err;
+
+    /* Encapsulate */
+    if (EVP_PKEY_encapsulate(ctx, wrapped, &wrappedlen, secret, &secretlen) <= 0)
+        goto err;
+
+    /* Compare outputs */
+    OSSL_SELF_TEST_oncorrupt_byte(st, wrapped);
+    if (wrappedlen != t->cipher_text_len
+            || memcmp(wrapped, t->cipher_text, t->cipher_text_len) != 0)
+        goto err;
+
+    OSSL_SELF_TEST_oncorrupt_byte(st, secret);
+    if (secretlen != t->secret_len
+            || memcmp(secret, t->secret, t->secret_len) != 0)
+        goto err;
+
+    ret = 1;
+ err:
+    OPENSSL_free(wrapped);
+    OPENSSL_free(secret);
+    EVP_PKEY_CTX_free(ctx);
+    OSSL_SELF_TEST_onend(st, ret);
+    return ret;
+}
+
+/*
+ * FIPS 140-3 IG 10.3.A resolution 14 mandates a CAST for ML-KEM
+ * decapsulation both for the rejection path and the normal path.
+ */
+static int self_test_kem_decapsulate(const ST_KAT_KEM *t, OSSL_SELF_TEST *st,
+                                     OSSL_LIB_CTX *libctx, EVP_PKEY *pkey,
+                                     int reject)
+{
+    int ret = 0;
+    EVP_PKEY_CTX *ctx = NULL;
+    unsigned char *secret = NULL, *alloced = NULL;
+    const unsigned char *cipher_text = t->cipher_text;
+    size_t secretlen;
+
+    OSSL_SELF_TEST_onbegin(st, OSSL_SELF_TEST_TYPE_KAT_KEM,
+                           self_test_kem_description(t, reject ? "decap reject"
+                                                               : "decap"));
+
+    if (reject) {
+        cipher_text = alloced = OPENSSL_memdup(t->cipher_text,
+                                               t->cipher_text_len);
+        if (alloced == NULL)
+            goto err;
+        *alloced ^= 1;
```

Comment:
Maybe just comment here that you are corrupting the first byte

Reply:
This is not the correct test. Instead, the test should take a known "corrupted" ciphertext and ensure that the output is a hash of that ciphertext with FO implicit rejection secret `z`. There are stock KATs for that.

Reply:
Where are these KATs?  I'm not finding them.